### DISCUSSION TITLE
Is it safe to use both the pg and the supabase-js client?

### DISCUSSION THREAD

**Discussion by jtlapp** [supabase/supabase#14226] [Questions] [answered]
I'd like to use [Kysely](https://kysely.dev/) with Supabase but also use the supabase-js client for user authentication and real-time. Kysely uses the pg client. I plan to deploy to Vercel.

Is it safe to be using both at the same time?

Will I be able to take advantage of pgBoucer?

Will I be able to implement a real-time chat solution?

Kysely only support's pg connection pools (via `Pool`) and not pg single-connection clients (via `Client`). Will it suffice to set the max pool size to 1?

**Comment by GaryAustin1** [comment, marked as answer]
You can use both a direct connection and a client.
This might be of interest... https://supabase.com/docs/guides/functions/kysely-postgres
I can't answer your Kysely questions, hopefully that example does.

What does implement a real-time chat solution mean?  If you are using the client then you can use realtime with that and it will reflect changes made by either the client or your direct connection.

**Reply by jtlapp** [reply]
Thank you! I just wanted to be sure the two clients wouldn't step on each other. Glad they work well together.

My connection pool question was about the postgres client rather than Kysely, which provides a `pg` adapter. I'll set max connections to 1 as done in the Kysely/Deno link you gave.

BTW, aside from demonstrating using a pool of 1 connection, the link only shows how to tweak Kysely for use with Deno. The only tweak that might be relevant to Supabase is the removal of Kysely's support for postgres streaming.

**Reply by jtlapp** [reply]
Follow up: Will the direct connection also use pgBouncer to pool pg connections? I can't find this information anywhere.

**Reply by GaryAustin1** [reply]
You would have to connect to the bouncer port at 6543 and with the correct options.  In the dashboard database settings page you will find info for the regular port and the bouncer port.  I have no idea what Kysley needs or does.  I was mainly answering your can the client coexist and pointing to one Kysley SB reference I found.

**Reply by jtlapp** [reply]
Excellent. And just to confirm:

1. I can issue multi-query transactions on the direct port and expect them all to use the same pg connection.
2. Each query to the bouncer port potentially uses a different pg connection.

**Reply by jtlapp** [reply]
I want to get this right so I can write a Supabase adapter for Kysely.

**Reply by jtlapp** [reply]
After reading [this page](https://devcenter.heroku.com/articles/best-practices-pgbouncer-configuration) on pgBouncer configuration, I'm thinking that port 6543 can handle both pooling and transactions, if we set "pool mode" to "transaction". The only reason to use port 5432 is if you need to keep a connection live across multiple transactions.
